File: breadth-first-search/is-graph-bipartite/is-graph-bipartite.go
package isgraphbipartite

func isBipartite(graph [][]int) bool {
	n := len(graph)
	set := make([]int, n)

	for i := 0; i < n; i++ {
		if set[i] != 0 {
			continue
		}

		queue := []int{}

		queue = append(queue, i)
		set[i] = 1

		for len(queue) > 0 {
			cur := queue[0]
			queue = queue[1:]
			node := graph[cur]

			for _, num := range node {
				if set[num] == 0 {
					set[num] = -set[cur]
					queue = append(queue, num)
				} else if set[num] == set[cur] {
					return false
				}
			}
		}
	}
	return true
}

File: breadth-first-search/is-graph-bipartite/is-graph-bipartite.ts
function isBipartite(graph: number[][]): boolean {
  const n: number = graph.length;
  const set: number[] = new Array(n).fill(0);

  for (let i = 0; i < n; i++) {
    if (set[i] != 0) {
      continue;
    }

    const queue: number[] = [i];
    set[i] = 1;

    while (queue.length > 0) {
      const cur: number = queue.shift()!;
      const node: number[] = graph[cur];

      for (let j = 0; j < node.length; j++) {
        if (set[node[j]] == 0) {
          set[node[j]] = -set[cur];
          queue.push(node[j]);
        } else if (set[node[j]] == set[cur]) {
          return false;
        }
      }
    }
  }
  return true;
}
